pod/mpi-operator   automatic restart

kubectl     logs  -p   pod/mpi-operator-69fd88fcf5-dn6jx     -n mpi-operator

I1008 11:58:30.415090       1 server.go:71] KUBEFLOW_NAMESPACE not set, use default namespace
I1008 11:58:30.415497       1 server.go:76] Using cluster scoped operator
I1008 11:58:30.415517       1 server.go:82] [API Version: v1alpha2 Version: v0.1.0-alpha Git SHA: Not provided. Go Version: go1.12.7 Go OS/Arch: linux/amd64]
I1008 11:58:30.415528       1 server.go:85] Server options: &{Kubeconfig: MasterURL: KubectlDeliveryImage:mpioperator/kubectl-delivery:latest Threadiness:2 PrintVersion:false EnableGangScheduling:false Namespace:}
I1008 11:58:47.438971       1 event.go:221] Event(v1.ObjectReference{Kind:"Endpoints", Namespace:"default", Name:"mpi-operator", UID:"9a98e104-d133-11e9-9781-00163e10d679", APIVersion:"v1", ResourceVersion:"13038044", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' mpi-operator-69fd88fcf5-dn6jx_f1ff468f-e9c2-11e9-8b64-aa50813bfad3 became leader
I1008 11:58:47.439236       1 mpi_job_controller.go:204] Setting up event handlers
I1008 11:58:47.439374       1 mpi_job_controller.go:338] Starting MPIJob controller
I1008 11:58:47.439388       1 mpi_job_controller.go:341] Waiting for informer caches to sync
E1008 11:58:47.447989       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:522
/usr/local/go/src/runtime/panic.go:82
/usr/local/go/src/runtime/signal_unix.go:390
/go/src/github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2/default.go:58
/go/src/github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2/default.go:58
/go/src/github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2/zz_generated.defaults.go:35
/go/src/github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2/zz_generated.defaults.go:29
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/runtime/scheme.go:391
/go/src/github.com/kubeflow/mpi-operator/pkg/controllers/v1alpha2/mpi_job_controller.go:839
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache/controller.go:195
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:554
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:203
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546
/go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71
/usr/local/go/src/runtime/asm_amd64.s:1337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x288 pc=0x1121563]

goroutine 114 [running]:
github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x12a4f20, 0x226d330)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2.setDefaults_TypeLauncher(...)
        /go/src/github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2/default.go:35
github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2.SetDefaults_MPIJob(...)
        /go/src/github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2/default.go:58
github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2.SetObjectDefaults_MPIJob(...)
        /go/src/github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2/zz_generated.defaults.go:35
github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2.RegisterDefaults.func1(0x13f8480, 0xc0004d8600)
        /go/src/github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v1alpha2/zz_generated.defaults.go:29 +0x93
github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/runtime.(*Scheme).Default(0xc000243e30, 0x16102c0, 0xc0004d8600)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/runtime/scheme.go:391 +0xb7
github.com/kubeflow/mpi-operator/pkg/controllers/v1alpha2.(*MPIJobController).addMPIJob(0xc0005eec80, 0x13f8480, 0xc0004d8600)
        /go/src/github.com/kubeflow/mpi-operator/pkg/controllers/v1alpha2/mpi_job_controller.go:839 +0x70
github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache/controller.go:195
github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x42ce1c, 0xc0003fe5e8, 0x0)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:554 +0x26d
github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0xc0003e1e38, 0x42c92f, 0xc0002e6120)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:203 +0xde
github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548 +0x89
github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0003fe768)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0003e1f68, 0xdf8475800, 0x0, 0x42d101, 0xc00040e060)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xf8
github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc0004a4800)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546 +0x9c
github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc00035a5a0, 0xc0006ec000)
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
created by github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
        /go/src/github.com/kubeflow/mpi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62

--- a/pkg/apis/kubeflow/v1alpha2/default.go
+++ b/pkg/apis/kubeflow/v1alpha2/default.go
@@ -29,16 +29,18 @@ func addDefaultingFuncs(scheme *runtime.Scheme) error {
 	return RegisterDefaults(scheme)
 }
 
-// setDefaults_TypeLauncher sets the default value to launcher.
-func setDefaults_TypeLauncher(spec *common.ReplicaSpec) {
-	// Only a `RestartPolicy` equal to `Never` or `OnFailure` is allowed for `Job`.
-	if spec.RestartPolicy != common.RestartPolicyNever {
-		spec.RestartPolicy = common.RestartPolicyOnFailure
+// setDefaultsTypeLauncher sets the default value to launcher.
+func setDefaultsTypeLauncher(spec *common.ReplicaSpec) {
+	if spec != nil {
+		// Only a `RestartPolicy` equal to `Never` or `OnFailure` is allowed for `Job`.
+		if spec.RestartPolicy != common.RestartPolicyNever {
+			spec.RestartPolicy = common.RestartPolicyOnFailure
+		}
 	}
 }
 
-// setDefaults_TypeWorker sets the default value to worker.
-func setDefaults_TypeWorker(spec *common.ReplicaSpec) {
+// setDefaultsTypeWorker sets the default value to worker.
+func setDefaultsTypeWorker(spec *common.ReplicaSpec) {
 }
 
 func SetDefaults_MPIJob(mpiJob *MPIJob) {
@@ -55,9 +57,8 @@ func SetDefaults_MPIJob(mpiJob *MPIJob) {
 	}
 
 	// set default to Launcher
-	setDefaults_TypeLauncher(mpiJob.Spec.MPIReplicaSpecs[MPIReplicaTypeLauncher])
+	setDefaultsTypeLauncher(mpiJob.Spec.MPIReplicaSpecs[MPIReplicaTypeLauncher])
 
 	// set default to Worker
-	setDefaults_TypeWorker(mpiJob.Spec.MPIReplicaSpecs[MPIReplicaTypeWorker])
-
+	setDefaultsTypeWorker(mpiJob.Spec.MPIReplicaSpecs[MPIReplicaTypeWorker])
 }

